### DISCUSSION TITLE
Support for ImageVector

### DISCUSSION THREAD

**Discussion by yschimke** [coil-kt/coil#1348] [Ideas]
I'm curious if there is a reason that ImageVector is not supported?

https://github.com/coil-kt/coil/blob/70300e5909a561d7e1026c660b8f28be3f0051a6/coil-compose-base/src/main/java/coil/compose/AsyncImagePainter.kt#L388

Seems like a useful variant of the data type which is Any?.

**Comment by colinrtwhite** [comment]
It’s unsupported as you can render an `ImageVector` using `Image` directly.

**Reply by yschimke** [reply]
OK, makes sense.  We have a reusable Chip, that takes Any? as the image data. It seemed unfortunate that we need to put the if statement in our code.

https://github.com/google/horologist/pull/330/files#diff-7c9f44c939721cf07ca16348362c65198ae1a117ca16cdb55c30306af24e38e9

Is it a performance related thing? Using Image a vastly simpler codepath?

But not going to argue it, just curious.

**Reply by colinrtwhite** [reply]
I believe it was initially added when `coil-compose` was part of Accompanist. I think it was mostly to avoid folks accidentally attempting to load `ImageVector`/`ImageBitmap` as those will fail to load without a custom `Fetcher` (basic `ImageLoaders` only understand `VectorDrawable`/`Bitmap`/`BitmapDrawable`). I agree it’s a little awkward with the if/else, but the alternative is adding custom fetchers (and removing the `require`). I’m not 100% opposed to removing the `require`. What do you think?

**Reply by yschimke** [reply]
Your call.

I'd prefer the custom fetcher, so it really is Any?. But maybe that's an antipattern for performance, when those ImageVectors should almost always just be Image.

**Reply by colinrtwhite** [reply]
I think we should leave it as is then. Can always revisit it later if necessary!

**Reply by bpappin** [reply]
ImageVector is pretty vital to all compose. We don't use resources *at all* anymore, they are all converted to compose ImageVector.
I'd like this to be revived I think.

**Reply by colinrtwhite** [reply]
@bpappin What's the benefit to using `ImageVector` with Coil when it's already supported by the standard `Image` composable? Using Coil for this case introduces extra overhead.

**Reply by bpappin** [reply]
There isn't, I didn't realize there was an alternative component to do that already.

**Reply by bpappin** [reply]
> @bpappin What's the benefit to using `ImageVector` with Coil when it's already supported by the standard `Image` composable? Using Coil for this case introduces extra overhead.

Image vector is the composable image format. We don't use XML or bitmaps anymore unless we have no other option.

**Reply by yschimke** [reply]
For me two benefits

1) building library components it's nice to accept an image as a param and display it without special casing
2) coil has other infra, maybe some interceptors should be applied?

**Reply by colinrtwhite** [reply]
@bpappin `ImageVector` is just Compose's abstraction on top of vector data. If you want Coil to display this data you're better off passing a reference to where the data is located to Coil (e.g. Android resource ID, file path, etc.).

**Reply by colinrtwhite** [reply]
@yschimke Gotcha, but the convenience would come with multiple costs:

- One frame of delay every time the composable is recomposed as the `ImageVector` can't be memory cached (no suitable cache key). We'll have to return it from a `Fetcher`, which runs on a background dispatcher, each time it's loaded.
- `ImageVector`/`ImageBitmap` can't be rendered on a standard Android/Skia canvas - they can only be drawn with Compose's `DrawScope.drawImage`. This doesn't play well with Coil's `coil3.Image` abstraction and would required some hacks to work around that.

I believe strongly that we want to nudge users to use the standard `Image` composable as it's faster and lets us avoid hacky code we would have to add to Coil to support this use case.

**Reply by yschimke** [reply]
Yeah, I guess `genId` is internal otherwise might suffice?

And for drawing without the additional bitmap caching there is no public API.

```
                @Suppress("INVISIBLE_MEMBER", "INVISIBLE_REFERENCE")
                with(painter.vector.root) { myCanvas.drawScope.draw() }
```

Yep, I take your point.

**Reply by bpappin** [reply]
> @bpappin `ImageVector` is just Compose's abstraction on top of vector data. If you want Coil to display this data you're better off passing a reference to where the data is located to Coil (e.g. Android resource ID, file path, etc.).

Yes, I understand what it is.

There is no resource id, it's in code now.
It's a literal block of code that's compiled, so there is no external data to load.
Basically what you get in the icon pack.

It solves a lot of problems because you don't need to manage resources, especially in multiplatform.

We only use bitmaps for photos, usually loaded remotely. Any SVG or XML image vector can be converted to a compiled format.

**Reply by bpappin** [reply]
> @yschimke Gotcha, but the convenience would come with multiple costs:
> 
> - One frame of delay every time the composable is recomposed as the `ImageVector` can't be memory cached (no suitable cache key). We'll have to return it from a `Fetcher`, which runs on a background dispatcher, each time it's loaded.
> - `ImageVector`/`ImageBitmap` can't be rendered on a standard Android/Skia canvas - they can only be drawn with Compose's `DrawScope.drawImage`. This doesn't play well with Coil's `coil3.Image` abstraction and would required some hacks to work around that.
> 
> I believe strongly that we want to nudge users to use the standard `Image` composable as it's faster and lets us avoid hacky code we would have to add to Coil to support this use case.

I think the reason the comments are growing here is that Coil is operating very differently to the way we work with compose. In effect, we have to "add hacky code" into our composition.

It's like the outlier of the components in the composition, because it doesn't follow the pattern. It works, and it's a great library, but when I use it, I feel like I'm adding a legacy view component, I can't just write the composition naturally and I have to comb through documentation to figure out how to do something that should be simple (often processing an image I shouldn't need to process).

It works, and I'm sure everyone here is happy to have it, and maybe a shift in the way we think about how it works would improve it for composable layouts.

**Reply by colinrtwhite** [reply]
@yschimke That would probably work if it was public, but even then we would have one frame of delay on the first load as `AsyncImage` loads the image async. Also the vector would occupy space in the memory cache, which would be better reserved for images that need to be decoded and are expensive to recreate (e.g. rasterized images loaded from disk).

I don't think `AsyncImage` should seek to replace the standard `Image` composable in all cases. This is what the [official Compose docs](https://developer.android.com/develop/ui/compose/graphics/images/loading) indicates as well. Images that don't need to be loaded and only need to be drawn (e.g. `ImageBitmap`, `ImageVector`, any `Painter`) shouldn't use `AsyncImage`. There are use cases where `Image` works much better and this is one of them.
